symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/WitnessEstimation.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE DerivingStrategies   #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module ZkFold.Symbolic.Compiler.ArithmeticCircuit.WitnessEstimation where
+
+
+
+import           Control.Applicative                            ()
+import           Data.Functor.Rep                               (Rep)
+import           GHC.Generics                                   (Generic)
+import           GHC.Integer                                    (Integer)
+import           GHC.Natural                                    (Natural)
+import           Prelude                                        (Eq, error, ($), (.), (==))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Data.ByteString                    ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Var
+
+
+
+data UVar a i
+  = ConstUVar a
+  | LinUVar a (SysVar i) a
+  | More
+  deriving Generic
+
+instance FromConstant a (UVar a i) where
+    fromConstant = ConstUVar
+instance FromConstant Natural a => FromConstant Natural (UVar a i) where fromConstant = ConstUVar . fromConstant
+instance FromConstant Integer a => FromConstant Integer (UVar a i) where fromConstant = ConstUVar . fromConstant
+
+instance (MultiplicativeSemigroup a, Eq a, AdditiveMonoid a) => Scale a (UVar a i) where
+  scale k (ConstUVar c) = ConstUVar $ k * c
+  scale k (LinUVar a x b) = if k == zero
+    then ConstUVar zero
+    else LinUVar (k * a) x (k * b)
+  scale k More = if k == zero
+    then ConstUVar zero
+    else More
+
+instance Scale Natural (UVar a i) where scale k = scale k . fromConstant
```

Comment:
`fromConstant` ends up being applied to the `UVar` itself which would cause this function to loop. I think you meant `scale k = scale (fromConstant k :: a)`